### PR TITLE
Add lowering of GET_ENVIRONMENT_VARIABLE intrinsic

### PR DESCRIPTION
**Pull request by Sezoir**
This patch adds the lowering of the "GET_ENVIRONMENT_VARIABLE" intrinsic
to the backend runtime hook implemented in patches D111394 and D112698.
It also renames the `isPresent` lambda to `isAbsent` and moves it out to
its own function in `Command.cpp`. Corresponding comment fixes for this
are also modified.

### CODE REVIEW

File: flang/lib/Optimizer/Builder/Runtime/Command.cpp [hunk 1]
```diff
@@ -13,6 +13,18 @@
 
 using namespace Fortran::runtime;
 
+// Cetain runtime intrinsics should only be ran when select parameters of the
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Builder/Runtime/Command.cpp:16]: nit: s/ran/run/

File: flang/lib/Optimizer/Builder/Runtime/Command.cpp [hunk 2]
```diff
@@ -13,6 +13,18 @@
...
+// intrisic are supplied. In certain cases one of these parameter may not be
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Builder/Runtime/Command.cpp:17]: s/parameter/parameters/

File: flang/lib/Optimizer/Builder/Runtime/Command.cpp [hunk 3]
```diff
@@ -13,6 +13,18 @@
...
+// given, however we wish to run the intrinsic due to another required parameter
+// being supplied. In this case we assign the missing parameter to have an
+// "absent" value. This typically happens in IntrinsicCall.cpp. For this reason
+// we need extra indirection with `isAbsent` for testing whether a given
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Builder/Runtime/Command.cpp:21]: Can you rephrase to not us the pronoun "we" in this comment?

**Review by schweitzpgi — COMMENTED** [review]
LGTM.

Make sure to rebase this to tip and recheck your tests. The format of FIR has changed. :)

File: flang/lib/Optimizer/Builder/Runtime/Command.cpp (hunk 1 above)
Comment (Leporacanthicus) [reply flang/lib/Optimizer/Builder/Runtime/Command.cpp:16]: Also s/cetain/certain/

Comment (Sezoir) [reply flang/lib/Optimizer/Builder/Runtime/Command.cpp]: Fixed.

File: flang/lib/Optimizer/Builder/Runtime/Command.cpp (hunk 2 above)
Comment (Sezoir) [reply flang/lib/Optimizer/Builder/Runtime/Command.cpp]: Fixed.

File: flang/lib/Optimizer/Builder/Runtime/Command.cpp (hunk 3 above)
Comment (Sezoir) [reply flang/lib/Optimizer/Builder/Runtime/Command.cpp]: Done.

File: flang/test/Lower/intrinsic-procedures/get_environment_variable.f90 [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+end subroutine name_and_value_only
+
+! CHECK-LABEL: func @_QPall_arguments(
+! CHECK-SAME: %[[nameArg:[^:]*]]: !fir.boxchar<1> {fir.bindc_name = "name"},
+! CHECK-SAME: %[[valueArg:.*]]: !fir.boxchar<1> {fir.bindc_name = "value"},
+! CHECK-SAME: %[[lengthArg:[^:]*]]: !fir.ref<i[[DEFAULT_INTEGER_SIZE]]> {fir.bindc_name = "length"},
+! CHECK-SAME: %[[statusArg:.*]]: !fir.ref<i[[DEFAULT_INTEGER_SIZE]]> {fir.bindc_name = "status"},
+! CHECK-SAME: %[[trimNameArg:.*]]: !fir.ref<!fir.logical<4>> {fir.bindc_name = "trim_name"},
+! CHECK-SAME: %[[errmsgArg:.*]]: !fir.boxchar<1> {fir.bindc_name = "errmsg"}) {
+subroutine all_arguments(name, value, length, status, trim_name, errmsg)
```
Comment (banach-space) [inline flang/test/Lower/intrinsic-procedures/get_environment_variable.f90:44]: [nit] You could move this case to the very end - as the most complex one and that differs from the others the most.

File: flang/test/Lower/intrinsic-procedures/get_environment_variable.f90 [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+! CHECK-LABEL: func @_QPname_and_trim_name_only(
+! CHECK-SAME: %[[nameArg:.*]]: !fir.boxchar<1> {fir.bindc_name = "name"},
+! CHECK-SAME: %[[trimNameArg:.*]]: !fir.ref<!fir.logical<4>> {fir.bindc_name = "trim_name"}) {
+subroutine name_and_trim_name_only(name, trim_name)
+    character(len=32) :: name
+    logical :: trim_name
+    call get_environment_variable(name, TRIM_NAME=trim_name)
+    ! CHECK-NOT: fir.call @_FortranAEnvVariableValue
+    ! CHECK-NOT: fir.call @_FortranAEnvVariableLength
+    ! CHECK-NEXT: return
```
Comment (banach-space) [inline flang/test/Lower/intrinsic-procedures/get_environment_variable.f90:134]: Could you add `! CHECK-NEXT: return` to all cases?

Comment (Sezoir) [reply flang/test/Lower/intrinsic-procedures/get_environment_variable.f90:93]: Done.

File: flang/test/Lower/intrinsic-procedures/get_environment_variable.f90 (hunk 4 above)
Comment (Sezoir) [reply flang/test/Lower/intrinsic-procedures/get_environment_variable.f90]: Done.

File: flang/lib/Optimizer/Builder/Runtime/Command.cpp [hunk 6]
```diff
@@ -68,3 +72,57 @@ void fir::runtime::genGetCommandArgument(fir::FirOpBuilder &builder,
...
+
+  // Only save result of `EnvVariableValue` if STATUS parameter provided
+  if (status) {
+    const mlir::Value statusLoaded = builder.create<fir::LoadOp>(loc, status);
+    mlir::Value resultCast =
+        builder.createConvert(loc, statusLoaded.getType(), valueResult);
+    builder.create<fir::StoreOp>(loc, resultCast, status);
+  }
+
+  // Similarly only run `EnvVariableLength` intrinsic if LENGTH parameter
```
[the remark this answers is not in the document]
Comment (Sezoir) [reply flang/lib/Optimizer/Builder/Runtime/Command.cpp]: Done.

**Comment by banach-space** [comment]
LGTM, thanks for addressing my comments!

**Review by banach-space — APPROVED** [review]

**Review by schweitzpgi — APPROVED** [review]
